File: app/src/main/java/org/worshipsongs/fragment/ServicesFragment.java
package org.worshipsongs.fragment;


import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import org.worshipsongs.CommonConstants;
import org.worshipsongs.R;
import org.worshipsongs.activity.FavouriteSongsActivity;
import org.worshipsongs.adapter.TitleAdapter;
import org.worshipsongs.listener.SongContentViewListener;
import org.worshipsongs.registry.ITabFragment;
import org.worshipsongs.service.FavouriteService;
import org.worshipsongs.service.PopupMenuService;
import org.worshipsongs.utils.CommonUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Author : Madasamy
 * Version : 3.x
 */

public class ServicesFragment extends Fragment implements TitleAdapter.TitleAdapterListener<String>, AlertDialogFragment.DialogListener, ITabFragment
{
    private FavouriteService favouriteService = new FavouriteService();
    private List<String> services = new ArrayList<>();
    private Parcelable state;
    private ListView serviceListView;
    private TitleAdapter<String> titleAdapter;
    private TextView infoTextView;
    private PopupMenuService popupMenuService = new PopupMenuService();

    public static ServicesFragment newInstance()
    {
        return new ServicesFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
//        favouriteService = new FavouriteService();
        if (savedInstanceState != null) {
            state = savedInstanceState.getParcelable(CommonConstants.STATE_KEY);
        }
        setHasOptionsMenu(true);
        initSetUp();
    }

    private void initSetUp()
    {
        services = favouriteService.findNames();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState)
    {
        View view = inflater.inflate(R.layout.songs_layout, container, false);
        setInfoTextView(view);
        setListView(view);
        return view;
    }

    private void setInfoTextView(View view)
    {
        infoTextView = (TextView) view.findViewById(R.id.info_text_view);
        infoTextView.setText(getString(R.string.favourite_info_message_));
        infoTextView.setLineSpacing(0, 1.2f);
        infoTextView.setVisibility(services.isEmpty() ? View.VISIBLE : View.GONE);
    }

    private void setListView(View view)
    {
        serviceListView = (ListView) view.findViewById(R.id.song_list_view);
        titleAdapter = new TitleAdapter<String>((AppCompatActivity) getActivity(), R.layout.songs_layout);
        titleAdapter.setTitleAdapterListener(this);
        titleAdapter.addObjects(services);
        serviceListView.setAdapter(titleAdapter);
    }

    @Override
    public void onSaveInstanceState(Bundle outState)
    {
        if (this.isAdded()) {
            outState.putParcelable(CommonConstants.STATE_KEY, serviceListView.onSaveInstanceState());
        }
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onResume()
    {
        super.onResume();
        initSetUp();
        refreshListView();
    }

    //Adapter listener methods
    @Override
    public void setViews(Map<String, Object> objects, final String text)
    {
        TextView titleTextView = (TextView) objects.get(CommonConstants.TITLE_KEY);
        titleTextView.setText(text);
        titleTextView.setOnLongClickListener(new TextViewLongClickListener(text));
        titleTextView.setOnClickListener(new TextViewOnClickListener(text));

        ImageView optionsImageView = (ImageView) objects.get(CommonConstants.OPTIONS_IMAGE_KEY);
        optionsImageView.setVisibility(View.VISIBLE);
        optionsImageView.setOnClickListener(new View.OnClickListener()
        {
            @Override
            public void onClick(View view)
            {
                popupMenuService.shareFavouritesInSocialMedia(view, text);
            }
        });
    }


    //Dialog Listener method
    @Override
    public void onClickPositiveButton(Bundle bundle, String tag)
    {
        String favouriteName = bundle.getString(CommonConstants.NAME_KEY, "");
        favouriteService.remove(favouriteName);
        services.clear();
        initSetUp();
        titleAdapter.addObjects(services);
        infoTextView.setVisibility(services.isEmpty() ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onClickNegativeButton()
    {
        // Do nothing
    }

    //Tab choices and reorder methods
    @Override
    public int defaultSortOrder()
    {
        return 4;
    }

    @Override
    public String getTitle()
    {
        return "playlists";
    }

    @Override
    public boolean checked()
    {
        return true;
    }

    @Override
    public void setListenerAndBundle(SongContentViewListener songContentViewListener, Bundle bundle)
    {

    }

    private class TextViewLongClickListener implements View.OnLongClickListener
    {
        private String serviceName;

        TextViewLongClickListener(String serviceName)
        {
            this.serviceName = serviceName;
        }

        @Override
        public boolean onLongClick(View v)
        {
            Bundle bundle = new Bundle();
            bundle.putString(CommonConstants.TITLE_KEY, getString(R.string.delete));
            bundle.putString(CommonConstants.MESSAGE_KEY, getString(R.string.message_delete_playlist, serviceName));
            bundle.putString(CommonConstants.NAME_KEY, serviceName);
            AlertDialogFragment alertDialogFragment = AlertDialogFragment.newInstance(bundle);
            alertDialogFragment.setDialogListener(ServicesFragment.this);
            alertDialogFragment.show(getActivity().getFragmentManager(), "DeleteConfirmationDialog");
            return false;
        }
    }

    private class TextViewOnClickListener implements View.OnClickListener
    {
        private String serviceName;

        TextViewOnClickListener(String serviceName)
        {
            this.serviceName = serviceName;
        }

        @Override
        public void onClick(View v)
        {
            Intent intent = new Intent(getActivity(), FavouriteSongsActivity.class);
            intent.putExtra(CommonConstants.SERVICE_NAME_KEY, serviceName);
            startActivity(intent);
        }
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser)
    {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisibleToUser) {
            hideKeyboard();
            initSetUp();
            refreshListView();
        }
    }

    private void hideKeyboard()
    {
        if (getActivity() != null) {
            CommonUtils.hideKeyboard(getActivity());
        }
    }

    private void refreshListView()
    {
        if (state != null) {
            serviceListView.onRestoreInstanceState(state);
        } else if (titleAdapter != null) {
            titleAdapter.addObjects(services);
            infoTextView.setVisibility(services.isEmpty() ? View.VISIBLE : View.GONE);
        }
    }
}

File: app/src/main/java/org/worshipsongs/fragment/ServicesFragment.kt
package org.worshipsongs.fragment


import android.app.Activity
import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment

import org.worshipsongs.CommonConstants
import org.worshipsongs.R
import org.worshipsongs.activity.FavouriteSongsActivity
import org.worshipsongs.adapter.TitleAdapter
import org.worshipsongs.listener.SongContentViewListener
import org.worshipsongs.registry.ITabFragment
import org.worshipsongs.service.FavouriteService
import org.worshipsongs.service.PopupMenuService
import org.worshipsongs.utils.CommonUtils

import java.util.ArrayList

/**
 * @author : Madasamy
 * @since : 3.x
 */

class ServicesFragment : Fragment(), TitleAdapter.TitleAdapterListener<String>, AlertDialogFragment.DialogListener, ITabFragment
{
    private val favouriteService = FavouriteService()
    private var services: MutableList<String> = ArrayList()
    private var state: Parcelable? = null
    private var serviceListView: ListView? = null
    private var titleAdapter: TitleAdapter<String>? = null
    private var infoTextView: TextView? = null
    private val popupMenuService = PopupMenuService()

    private val negativeButtonListener: DialogInterface.OnClickListener
        get() = DialogInterface.OnClickListener { dialog, which -> dialog.dismiss() }

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        if (savedInstanceState != null)
        {
            state = savedInstanceState.getParcelable(CommonConstants.STATE_KEY)
        }
        setHasOptionsMenu(true)
        initSetUp()
    }

    private fun initSetUp()
    {
        services = favouriteService.findNames()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
    {
        val view = inflater.inflate(R.layout.songs_layout, container, false)
        setInfoTextView(view)
        setListView(view)
        return view
    }

    private fun setInfoTextView(view: View)
    {
        infoTextView = view.findViewById<View>(R.id.info_text_view) as TextView
        infoTextView!!.text = getString(R.string.favourite_info_message_)
        infoTextView!!.setLineSpacing(0f, 1.2f)
        infoTextView!!.visibility = if (services.isEmpty()) View.VISIBLE else View.GONE
    }

    private fun setListView(view: View)
    {
        serviceListView = view.findViewById<View>(R.id.song_list_view) as ListView
        titleAdapter = TitleAdapter((activity as AppCompatActivity?)!!, R.layout.songs_layout)
        titleAdapter!!.setTitleAdapterListener(this)
        titleAdapter!!.addObjects(services)
        serviceListView!!.adapter = titleAdapter
    }

    override fun onSaveInstanceState(outState: Bundle)
    {
        if (this.isAdded)
        {
            outState.putParcelable(CommonConstants.STATE_KEY, serviceListView!!.onSaveInstanceState())
        }
        super.onSaveInstanceState(outState)
    }

    override fun onResume()
    {
        super.onResume()
        initSetUp()
        refreshListView()
    }

    //Adapter listener methods
   override fun setViews(objects: Map<String, Any>, text: String?)
    {
        val titleTextView = objects[CommonConstants.TITLE_KEY] as TextView?
        titleTextView!!.text = text!!
        titleTextView.setOnLongClickListener(TextViewLongClickListener(text))
        titleTextView.setOnClickListener(TextViewOnClickListener(text))

        val optionsImageView = objects[CommonConstants.OPTIONS_IMAGE_KEY] as ImageView?
        optionsImageView!!.visibility = View.VISIBLE
        optionsImageView.setOnClickListener { view -> popupMenuService.shareFavouritesInSocialMedia(activity as Activity, view, text) }
    }


    //Dialog Listener method
    override fun onClickPositiveButton(bundle: Bundle?, tag: String?)
    {
        val favouriteName = bundle!!.getString(CommonConstants.NAME_KEY, "")
        favouriteService.remove(favouriteName)
        services.clear()
        initSetUp()
        titleAdapter!!.addObjects(services)
        infoTextView!!.visibility = if (services.isEmpty()) View.VISIBLE else View.GONE
    }

    override fun onClickNegativeButton()
    {
        // Do nothing
    }

    //Tab choices and reorder methods
    override fun defaultSortOrder(): Int
    {
        return 4
    }

    override val title: String get() { return "playlists"}


    override fun checked(): Boolean
    {
        return true
    }

    override fun setListenerAndBundle(songContentViewListener: SongContentViewListener?, bundle: Bundle)
    {

    }

    private inner class TextViewLongClickListener internal constructor(private val serviceName: String) : View.OnLongClickListener
    {

        private val positiveButtonListener: DialogInterface.OnClickListener
            get() = DialogInterface.OnClickListener { dialog, which ->
                favouriteService.remove(serviceName)
                services.clear()
                initSetUp()
                titleAdapter!!.addObjects(services)
                infoTextView!!.visibility = if (services.isEmpty()) View.VISIBLE else View.GONE
            }

        override fun onLongClick(v: View): Boolean
        {
            val alertDialogBuilder = AlertDialog.Builder(activity)
            alertDialogBuilder.setTitle(getString(R.string.delete))
            alertDialogBuilder.setMessage(getString(R.string.message_delete_playlist, serviceName))
            alertDialogBuilder.setPositiveButton(getString(R.string.ok), positiveButtonListener)
            alertDialogBuilder.setNegativeButton(R.string.cancel, negativeButtonListener)
            alertDialogBuilder.show()
            return false
        }
    }

    private inner class TextViewOnClickListener internal constructor(private val serviceName: String) : View.OnClickListener
    {

        override fun onClick(v: View)
        {
            val intent = Intent(activity, FavouriteSongsActivity::class.java)
            intent.putExtra(CommonConstants.SERVICE_NAME_KEY, serviceName)
            startActivity(intent)
        }
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean)
    {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser)
        {
            hideKeyboard()
            initSetUp()
            refreshListView()
        }
    }

    private fun hideKeyboard()
    {
        if (activity != null)
        {
            CommonUtils.hideKeyboard(activity)
        }
    }

    private fun refreshListView()
    {
        if (state != null)
        {
            serviceListView!!.onRestoreInstanceState(state)
        } else if (titleAdapter != null)
        {
            titleAdapter!!.addObjects(services)
            infoTextView!!.visibility = if (services.isEmpty()) View.VISIBLE else View.GONE
        }
    }

    companion object
    {

        fun newInstance(): ServicesFragment
        {
            return ServicesFragment()
        }
    }
}
